src/QsCompiler/CompilerOptimizations/Utils.fs

```diff
@@ -69,88 +31,34 @@ let internal rangeLiteralToSeq (r: Expr): seq<int64> =
-        )) + "]"
-
-
-/// Represents a (possibly-nested) tuple of strings.
-/// Used as a way to compare QsTuples and SymbolTuples.
-type internal StringTuple =
-| SingleItem of string
-| MultipleItems of seq<StringTuple>
-
-    static member fromQsTuple (x: QsTuple<LocalVariableDeclaration<QsLocalSymbol>>): StringTuple =
-        match x with
-        | QsTupleItem item ->
-            match item.VariableName with
-            | ValidName n -> SingleItem n.Value
-            | InvalidName -> SingleItem "__invalid__"
-        | QsTuple items -> MultipleItems (Seq.map StringTuple.fromQsTuple items)
-
-    static member fromSymbolTuple (x: SymbolTuple): StringTuple =
-        match x with
-        | InvalidItem -> SingleItem "__invalid__"
-        | VariableName n -> SingleItem n.Value
-        | VariableNameTuple items -> MultipleItems (Seq.map StringTuple.fromSymbolTuple items)
-        | DiscardedItem -> SingleItem "__discarded__"
-
-    override this.ToString() =
-        match this with
-        | SingleItem item -> item
-        | MultipleItems items -> "(" + String.Join(", ", items) + ")"
-
-
-/// Modifies the VariablesDict by setting the given argument tuple to the given values
-let rec internal fillVars (vars: VariablesDict) (argTuple: StringTuple, arg: Expr): unit =
-    match argTuple with
-    | SingleItem item -> vars.defineVar(item, arg)
-    | MultipleItems items ->
-        match arg with
-        | ValueTuple vt ->
-            QsCompilerError.Verify (
-                vt.Length = Seq.length items,
-                "Matched tuples must have the same length")
-            vt |> Seq.map (fun x -> x.Expression) |>
-            Seq.zip items |> Seq.map (fillVars vars) |> List.ofSeq |> ignore
-        | _ ->
-            QsCompilerError.Verify (
-                1 = Seq.length items,
-                "Expecting a single-element tuple")
-            Seq.zip items [arg] |> Seq.map (fillVars vars) |> List.ofSeq |> ignore
+/// Converts a QsTuple to a SymbolTuple
+let rec internal toSymbolTuple (x: QsTuple<LocalVariableDeclaration<QsLocalSymbol>>): SymbolTuple =
+    match x with
+    | QsTupleItem item ->
+        match item.VariableName with
+        | ValidName n -> VariableName n
+        | InvalidName -> InvalidItem
+    | QsTuple items when items.Length = 1 ->
+        toSymbolTuple items.[0]
+    | QsTuple items ->
+        VariableNameTuple ((Seq.map toSymbolTuple items).ToImmutableArray())
+
+/// Wraps a QsExpressionType in a basic TypedExpression
+/// The returned TypedExpression has no type param / inferred info / range information,
+/// and it should not be used for any code step that requires this information.
+let internal wrapExpr (expr: Expr) (bt: TypeKind): TypedExpression =
+    let ii = {IsMutable=false; HasLocalQuantumDependency=false}
+    TypedExpression.New (expr, ImmutableDictionary.Empty, ResolvedType.New bt, ii, Null)
+
+/// Wraps a QsStatementKind in a basic QsStatement
```

Comment:
What does "basic" mean?

Reply:
It means that I'm leaving out all the optional information, so that I'm returning close to a bare wrapper around a StatementKind. I'll add more comments in a future PR.

Reply:
Yes, the adding more to the doc comment was where I was going. :)